core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java

```diff
@@ -0,0 +1,313 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Multimap;
+import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import io.trino.client.NodeVersion;
+import io.trino.cost.StatsAndCosts;
+import io.trino.execution.ExecutionFailureInfo;
+import io.trino.execution.NodeTaskMap;
+import io.trino.execution.RemoteTask;
+import io.trino.execution.StageId;
+import io.trino.execution.StateMachine;
+import io.trino.execution.TaskId;
+import io.trino.execution.TaskState;
+import io.trino.execution.TaskStatus;
+import io.trino.execution.buffer.OutputBufferStatus;
+import io.trino.metadata.InMemoryNodeManager;
+import io.trino.metadata.InternalNode;
+import io.trino.metadata.Split;
+import io.trino.sql.planner.Partitioning;
+import io.trino.sql.planner.PartitioningScheme;
+import io.trino.sql.planner.PlanFragment;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.plan.PlanFragmentId;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.testing.TestingMetadata;
+import io.trino.util.FinalizerService;
```

Comment:
1. Can the integration test have multiple workers? If so, how can it check the number of workers?
2. The feature is limited to the scheduler, not connector. So, I think unit tests suffice.

Reply:
> Can the integration test have multiple workers? If so, how can it check the number of workers?

Yes. ptal at `testScaleWriters`. It's testing scaling of writers with what is now `isAverageBufferUtilizationFull`.

> The feature is limited to the scheduler, not connector. So, I think unit tests suffice.

I think it's more than that. For weighted scaling to work multiple components need to cooperate, coordinator is just a tip. It would be better to have an e2e test.
Given that we already have `testScaleWriters`, it should be relatively easy to add a skewed test (e.g. with skewed join before insert). In `BaseHiveConnectorTest` by default there are 3 workers.

Reply:
Let me try. But, can we do that in a follow-up PR?